File: river/stats/maximum.py
from __future__ import annotations

import math

from river import stats, utils


class Max(stats.base.Univariate):
    """Running max.

    Attributes
    ----------
    max : float
        The current max.

    Examples
    --------

    >>> from river import stats

    >>> X = [1, -4, 3, -2, 5, -6]
    >>> maximum = stats.Max()
    >>> for x in X:
    ...     maximum.update(x)
    ...     print(maximum.get())
    1
    1
    3
    3
    5
    5

    """

    def __init__(self) -> None:
        self.max: float = -math.inf

    def update(self, x: float) -> None:
        if x > self.max:
            self.max = x

    def get(self) -> float:
        return self.max


class RollingMax(stats.base.RollingUnivariate):
    """Running max over a window.

    Parameters
    ----------
    window_size
        Size of the rolling window.

    Examples
    --------

    >>> from river import stats

    >>> X = [1, -4, 3, -2, 2, 1]
    >>> rolling_max = stats.RollingMax(window_size=2)
    >>> for x in X:
    ...     rolling_max.update(x)
    ...     print(rolling_max.get())
    1
    1
    3
    3
    2
    2

    """

    def __init__(self, window_size: int) -> None:
        self.window: utils.SortedWindow[float] = utils.SortedWindow(size=window_size)

    @property
    def window_size(self) -> int:
        return self.window.size

    def update(self, x: float) -> None:
        self.window.append(x)

    def get(self) -> float:
        try:
            return self.window[-1]
        except IndexError:
            raise stats.NotEnoughSamples


class AbsMax(stats.base.Univariate):
    """Running absolute max.

    Attributes
    ----------
    abs_max : float
        The current absolute max.

    Examples
    --------

    >>> from river import stats

    >>> X = [1, -4, 3, -2, 5, -6]
    >>> abs_max = stats.AbsMax()
    >>> for x in X:
    ...     abs_max.update(x)
    ...     print(abs_max.get())
    1
    4
    4
    4
    5
    6

    """

    def __init__(self) -> None:
        self.abs_max: float = 0.0

    def update(self, x: float) -> None:
        if abs(x) > self.abs_max:
            self.abs_max = abs(x)

    def get(self) -> float:
        return self.abs_max


class RollingAbsMax(stats.base.RollingUnivariate):
    """Running absolute max over a window.

    Parameters
    ----------
    window_size
        Size of the rolling window.

    Examples
    --------

    >>> from river import stats

    >>> X = [1, -4, 3, -2, 2, 1]
    >>> rolling_absmax = stats.RollingAbsMax(window_size=2)
    >>> for x in X:
    ...     rolling_absmax.update(x)
    ...     print(rolling_absmax.get())
    1
    4
    4
    3
    2
    2

    """

    def __init__(self, window_size: int) -> None:
        self.window: utils.SortedWindow[float] = utils.SortedWindow(size=window_size)

    @property
    def window_size(self) -> int:
        return self.window.size

    def update(self, x: float) -> None:
        self.window.append(abs(x))

    def get(self) -> float:
        try:
            return self.window[-1]
        except IndexError:
            raise stats.NotEnoughSamples

File: rust_src/maximum.rs
use crate::sorted_window::SortedWindow;
use crate::stats::Univariate;
use num::{Float, FromPrimitive};
use serde::{Deserialize, Serialize};
use std::ops::{AddAssign, SubAssign};
/// Running max.
/// # Examples
/// ```
/// use watermill::maximum::Max;
/// use watermill::stats::Univariate;
/// let mut running_max: Max<f64> = Max::new();
/// for i in 1..10{
///     running_max.update(i as f64);
/// }
/// assert_eq!(running_max.get(), 9.0);
/// ```
///
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
pub struct Max<F: Float + FromPrimitive + AddAssign + SubAssign> {
    pub max: F,
}
impl<F: Float + FromPrimitive + AddAssign + SubAssign> Default for Max<F> {
    fn default() -> Self {
        Self {
            max: F::min_value(),
        }
    }
}
impl<F: Float + FromPrimitive + AddAssign + SubAssign> Max<F> {
    pub fn new() -> Self {
        Self::default()
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Univariate<F> for Max<F> {
    fn update(&mut self, x: F) {
        if self.max < x {
            self.max = x;
        }
    }
    fn get(&self) -> F {
        self.max
    }
}

/// Running absolute max.
/// # Examples
/// ```
/// use watermill::maximum::AbsMax;
/// use watermill::stats::Univariate;
/// let mut running_abs_max: AbsMax<f64> = AbsMax::new();
/// for i in -17..10{
///     running_abs_max.update(i as f64);
/// }
/// assert_eq!(running_abs_max.get(), 17.0);
/// ```
///
#[derive(Debug)]
pub struct AbsMax<F: Float + FromPrimitive + AddAssign + SubAssign> {
    abs_max: F,
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Default for AbsMax<F> {
    fn default() -> Self {
        Self {
            abs_max: F::from_f64(0.0).unwrap(),
        }
    }
}
impl<F: Float + FromPrimitive + AddAssign + SubAssign> AbsMax<F> {
    pub fn new() -> Self {
        Self::default()
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Univariate<F> for AbsMax<F> {
    fn update(&mut self, x: F) {
        if self.abs_max < x.abs() {
            self.abs_max = x.abs();
        }
    }
    fn get(&self) -> F {
        self.abs_max
    }
}

/// Rolling max.
/// # Arguments
/// * `window_size` - Size of the rolling window.
/// # Examples
/// ```
/// use watermill::maximum::RollingMax;
/// use watermill::stats::Univariate;
/// let mut rolling_max: RollingMax<f64> = RollingMax::new(3);
/// for i in 1..10{
///     rolling_max.update(i as f64);
/// }
/// assert_eq!(rolling_max.get(), 9.0);
/// ```
///
#[derive(Serialize, Deserialize)]
pub struct RollingMax<F: Float + FromPrimitive + AddAssign + SubAssign> {
    sorted_window: SortedWindow<F>,
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> RollingMax<F> {
    pub fn new(window_size: usize) -> Self {
        Self {
            sorted_window: SortedWindow::new(window_size),
        }
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Univariate<F> for RollingMax<F> {
    fn update(&mut self, x: F) {
        self.sorted_window.push_back(x);
    }
    fn get(&self) -> F {
        self.sorted_window.back()
    }
}
